opm/simulators/wells/StandardWell_impl.hpp

```diff
@@ -580,14 +600,29 @@ namespace Opm
                     const unsigned gasCompIdx = Indices::canonicalToActiveComponentIndex(FluidSystem::gasCompIdx);
                     // q_os = q_or * b_o + rv * q_gr * b_g
                     // q_gs = q_gr * g_g + rs * q_or * b_o
+                    // q_gr = 1 / (b_g * d) * (q_gs - rs * q_os)
                     // d = 1.0 - rs * rv
                     const EvalWell d = this->extendEval(1.0 - fs.Rv() * fs.Rs());
-                    // q_gr = 1 / (b_g * d) * (q_gs - rs * q_os)
-                    if(FluidSystem::gasPhaseIdx == phaseIdx)
+                    if (d <= 0.0) {
+                        std::ostringstream sstr;
+                        sstr << "Problematic d value " << d << " obtained for well " << this->name()
+                            << " during calculateSinglePerf with rs " << fs.Rs()
+                            << ", rv " << fs.Rv()
+                            << " obtaining d " << d
+                            << " Continue as if no dissolution (rs = 0) and vaporization (rv = 0) "
+                            << " for this connection.";
+                        deferred_logger.debug(sstr.str());
+                    }
+                    if(FluidSystem::gasPhaseIdx == phaseIdx && d > 0.0) {
                         cq_r_thermal = (cq_s[gasCompIdx] - this->extendEval(fs.Rs()) * cq_s[oilCompIdx]) / (d * this->extendEval(fs.invB(phaseIdx)) );
                     // q_or = 1 / (b_o * d) * (q_os - rv * q_gs)
-                    if(FluidSystem::oilPhaseIdx == phaseIdx)
+                    } else if(FluidSystem::oilPhaseIdx == phaseIdx && d > 0.0) {
                         cq_r_thermal = (cq_s[oilCompIdx] - this->extendEval(fs.Rv()) * cq_s[gasCompIdx]) / (d * this->extendEval(fs.invB(phaseIdx)) );
+                    } else {
+                        assert(d <= 0.0);
+                        // for d <= 0.0 we continue as if rs = 0 and rv = 0
+                        cq_r_thermal = cq_s[activeCompIdx] / this->extendEval(fs.invB(phaseIdx));
```

Comment:
Nitpicking here. 

This line can moved to below the line 614 `                        deferred_logger.debug(sstr.str());`, and for the gasPhase and oilPhase if conditions do not need to `d > 0.0`.  And also we do not need the `assert`.